Add LogBook support to HomeKit

ehendrix23 opened the pull request:
## Description:
Initial set to add HomeKit entries to LogBook. Setting the framework for it.
Currently following logbook entries will be created:
-) When an accessory is added for HomeKit.
-) When a lock is opened or closed.

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**


If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

ehendrix23 committed:
HomeKit LogBook

Initial set to add HomeKit entries to LogBook. Setting the framework for it.
Currently following logbook entries will be created:
-) When an accessory is added for HomeKit.
-) When a lock is opened or closed.

ehendrix23 commented:
@cdce8p 
Here is the first try. Have a look on what I did. :-)

It is working on mine. When I start HASS I get entries in LogBook about the accessories that have been added to HomeKit. It provides the accessory type and the friendly name.
For example:
> HomeKit Add accessory Lock for Front Door

Then when I unlock (or lock) the door it puts in logbook an entry like:
> HomeKit Send command unlock for Front Door

Front Door here is my display name.

cdce8p commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +287,28 @@ def humanify(hass, events):
+                    'context_user_id': event.context.user_id
+                }
+
+
```
One empty line to much :)

ehendrix23 replied:
Not sure why this was not resolved as I did remove the extra line in there. Also checked the code with pylint, pydocstyle, and flake8 and no issues reported.
Will set this to resolve manually. :-)

cdce8p reviewed (changes requested):
Left a few comments. Mostly about unnecessary empty lines :slightly_smiling_face:
Personally I think we should move the `logbook` changes to a new PR, but we could leave them in there until the rest is done to make live testing easier.

ehendrix23 commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -168,8 +169,16 @@ def get_accessory(hass, driver, state, aid, config):
+
+    accessory = TYPES[a_type](hass, driver, name, state.entity_id, aid, config)
+    accessory.sent_logbook_message(
+        EVENT_HOMEKIT_CHANGED, message
```
I added this in there as I had a few times when I noticed something wasn't in Homekit I was expecting to be there. Normally speaking one would not run with debug and I had noticed that automations put in there an initial statement as well that they are now "active". 
This then allows an end-user to more easily see/conform that from HASS it is added as an accessory to HomeKit without having to go into log files and having to turn debug on first.

I personally like to be able to easily see in the logbook that the accessories have been added, and it is only once they go in there upon start hence it would not be spamming the logbook as such I think.

Let me know if you want to keep it there (my preference) or not, if not I'll just revert that piece of code. :-)

cdce8p replied:
Would you be ok with leaving it out for now and come to it later, once this is merged?
I would like to focus this PR one the `changed` messages, since this way it will already be quite large.

ehendrix23 replied:
Yep, removed that part.

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -9,13 +9,14 @@
     async_track_state_change, track_point_in_utc_time)
 from homeassistant.util import dt as dt_util
 
+
```
Wanted to be thorough. :-) Yeah, figured I'll clean it up once we have a final approach. :-)

cdce8p replied:
Quick tip: Don't know if you have seen this already: https://developers.home-assistant.io/docs/en/development_testing.html#testing-outside-of-tox

Especially the part about `script/lint`. I run it all the time when I'm working on something new.

ehendrix23 replied:
Yeah, right after I did the commit and put it in the PR and then saw the check items come back I knew I forgot something. :-)
Figured that instead of going back and fixing it right then to first get your initial feedback and then fix extra lines etc. together with changes based on your feedback. :-)

Ran it all this time around, so should come back good. :-)

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -25,6 +26,7 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+EVENT_HOMEKIT_CHANGED = 'homekit_state_change'
```
Totally agree, but I noticed that for Alexa it was not put in const.py but in the smart_home.py itself; hence followed that example. 
I'll move it into const.py instead.

ehendrix23 replied:
Moved into homekit/const.py

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +139,47 @@ def update_state(self, new_state):
+        })
+
+
+    def set_state(self, DOMAIN, service, params, logbook_message):
```
Yes, idea is that anytime something from HomeKit is called to change (i.e. lock/unlock, setting speed, turning on sprinkler, ...) this method is called then to put the information into logbook and then call the service to perform the action.
Will rename. :-)

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +139,47 @@ def update_state(self, new_state):
+                message = "Send command {}".format(service)
+
+            self.sent_logbook_message(EVENT_HOMEKIT_CHANGED, message)
+        except Exception as ex:
```
Figure that firing the event itself results into going through some additional coding itself. My idea here for the try/except is to ensure that if something goes wrong/fails for the logbook that it does not interfere with the action itself.

For example, I trigger the event to lock my front door and for some reason (whatever it might be) the coding for the logbook (be it what we have or firing) has any kind of issue then the actual event to lock the door is not performed. By putting it within a try/except we make sure that the event to lock the door will still occur. 
More of a safety thing to ensure logging entries in logbook do not interfere with calling the action in case of failure.

We could call the service 1st, but then in the logbook it might show the action first before the entry from HomeKit which would be confusing.

cdce8p replied:
The beauty of the Home Assistant core being `async` is that one part doesn't interface directly with another. In this case:
 1. The HomeKit component will just ask the core to fire the event
 2. The core will then add jobs to do all tasks that listen for the event to the `event loop`
 3. The HomeKit method runs until completion
 4. The event loop executes previously added jobs

In summary, if something fails in the logbook component, it would **not** affect `HomeKit`. The `try` block isn't needed.

If you'r interested in async, this might be a good start: https://developers.home-assistant.io/docs/en/asyncio_101.html

ehendrix23 replied:
Thanks for the link, will definitely read on it. My main point was just if there was anything going wrong on the firing of the event itself then it would not stop changing the state in itself.
As I mentioned, a safety precaution in case something happens. 

Removed the try/except.

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +139,47 @@ def update_state(self, new_state):
+
+        self.hass.bus.fire(homekit_event, {
+            ATTR_ENTITY_ID: self.entity_id,
+            'message': logbook_attr_message
```
I thought about that but figured that this way we define the message within HomeKit itself and only pass it on to LogBook. Means that in the future we can easily add/change/... the message without having to touch LogBook.
For example, message for turning on sprinkler could be like:
> Homekit send command to turn on sprinkler for front yard

and for unlocking a door:
> Homekit send command to unlock for front door

for changing fan direction:
> Homekit send command to reverse fan direction for family room

And any future homekit accessories can thus define it on what makes the most sense without then also having to go in and modify logbook coding itself.

ehendrix23 commented on homeassistant/components/homekit/type_locks.py:
```diff
@@ -54,7 +55,8 @@ def set_state(self, value):
         if self._code:
             params[ATTR_CODE] = self._code
-        self.hass.services.call(DOMAIN, service, params)
+        super().set_state(DOMAIN, service, params, service)
```
In this case it is "service" twice as for "locks" what is passed as the message is the same as what is in service. But for other accessories it would be different.
What I can do is make it optional and if not provided then use service instead.

cdce8p replied:
Optional is a good start. It depends how we will structure the `message` then, with which if clauses ;)

ehendrix23 replied:
Changed it to optional and if not provided then I will take what was provided from service instead.

ehendrix23 committed:
Updates based on comments

Updates performed based on review and comments.

ehendrix23 commented:
@cdce8p  Completed the changes, have a look. One item you marked for some reason did not get resolved/outdated. I checked logbook.py with flake8, pylint, and pydocstyle and no issues reported with the updated one.

cdce8p commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -169,6 +169,7 @@ def get_accessory(hass, driver, state, aid, config):
         return None
 
     _LOGGER.debug('Add "%s" as "%s"', state.entity_id, a_type)
+
```
Empty line :D

ehendrix23 replied:
Removed. :-)

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +138,35 @@ def update_state(self, new_state):
+        else:
+            message = "send command {}".format(service)
+
+        self.sent_logbook_message(EVENT_HOMEKIT_CHANGED, message)
```
Can you combine these two methods into `call_service`? I see no reason to have them sparate.

ehendrix23 replied:
I had them separate to allow other items to send a log message as well (i.e. adding an accessory).
Based on our conversations I believe that we will only create the logbook entry for state changes and anything else would be done by triggering an event and then the user would use automation rules to create the logbook entry. For example adding an accessory.
Combined it now and changed some other parts on it. Have a look. :-)

ehendrix23 replied:
On the logger debug statement. I did not move it to call_service right now. Currently the debug statement is the 1st statement being done, before anything else will be done. Hence if there is any kind of issue or so when going through the method the bug statement would have been written that it entered that method.
If we move it however to call_service then the debug statement will not be written till almost the end. Thus if there is a failure due to an unexpected value the debug statement would  not be written out then.

cdce8p reviewed (commented):
Already looking much better now!
Will do an in deep review tomorrow but wanted to leave you some quick comments.
Can you try moving the logger debug statment to call_service as well?

ehendrix23 committed:
Further updates based on review

Moved firing of event into call_service method, removing sent_logbook_message.
Based that logbook messages will only be done for requests from HomeKit, simplified message creation accordingly.
Added display_name to data for event.

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +138,24 @@ def update_state(self, new_state):
         """
         raise NotImplementedError()
 
+    def call_service(self, acc_domain, service, params, logbook_message=None):
```
I would like to change the parameter names here to
`def call_service(self, domain, service, service_data, attribute=None)`

`service_data` only in this method. We don't need to rename it in the `types`.

ehendrix23 replied:
Updated.

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +138,24 @@ def update_state(self, new_state):
+
+        Then call the service to change the state.
+        This method is to be called from methods that change states using
+        self.call_service(DOMAIN, service, params, logbook_message).
```
```py
"""Fire event and call service for changes from HomeKit."""
```
is enough.

ehendrix23 replied:
Updated.

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +138,24 @@ def update_state(self, new_state):
+        self.call_service(DOMAIN, service, params, logbook_message).
+        """
+        if logbook_message is None:
+            logbook_message = service
```
Construct the `event_data` dict beforehand and only add the `value` if given.

ehendrix23 replied:
Based on changes done do not have this anymore. Just firing event with the 4 parameters.

cdce8p replied:
The `value` parameter will still be optional, but I guess we could just pass `None` as well if we do the check and message construction in the `logbook` component.

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +138,24 @@ def update_state(self, new_state):
+
+        self.hass.bus.fire(EVENT_HOMEKIT_CHANGED, {
+            ATTR_ENTITY_ID: self.entity_id,
+            'display_name': self.display_name,
```
Maybe a new constant: `ATTR_DISPLAY_NAME`? We would need `ATTR_VALUE` as well. A new section between `Misc` and `Config` named `Attributes` would be optimal.

ehendrix23 replied:
These would go into the HomeKit constants then?

cdce8p replied:
:+1:

ehendrix23 replied:
Added const.py for HomeKit.

cdce8p commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +287,32 @@ def humanify(hass, events):
+                else:
+                    action = "send command"
+
+                if display_name:
```
The `display_name` will always be given, since the fallback is the `entity_id`. No need to check for it.

ehendrix23 replied:
Not completely following. Would you mind re-checking on what it looks like now (once I commit)?

Appreciate it!

cdce8p replied:
Will do.

What I was trying to say is that you will always have a `display_name`, so `if display_name` isn't necessary.

cdce8p commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +287,32 @@ def humanify(hass, events):
+
+                if message:
+                    action = "send command {}".format(message)
+                elif entity_id:
```
Instead of `if / elif / else` blocks. It could be useful to construct a block beforehand and just add it:
`msg_value = " to {}".format(value) if value else ''`

`message = "send command {}{} for {}".format(service, msg_value, display_name)`

ehendrix23 replied:
Updated

ehendrix23 committed:
Changed to remove message and use service&attribute

Changed for the log message now to be created using the service & attribute instead of providing a specific log message.
Added constants to homekit/const.
Updated parameter name acc_domain to domain in call_service.

cdce8p commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +288,29 @@ def humanify(hass, events):
+            elif event.event_type == EVENT_HOMEKIT_CHANGED:
+                data = event.data
+                entity_id = data.get(ATTR_ENTITY_ID)
+                dsp_name = data.get(ATTR_DISPLAY_NAME) or entity_id
```
Actually, instead of `data.get` just use `data[ATTR_ENTITY_ID]` and so on, since they are all passed in the event_data every time.

ehendrix23 replied:
My reasoning for this was if the event is sent from something else or so and not from call_service itself, hence did it this way from a safety perspective.
If you do not believe this is required and we assume that data is always there then I can process it accordingly. :-)

cdce8p replied:
Since this is a HomeKit only event, I think we'll be fine. We can even streamline it further and integrate those calls directly into the message construction (like Alexa is doing). No need to define a new variable for just one interaction. An exception for me would be the `entity_id` (if we want to be a crazy, it could even remain with `get` ;)

ehendrix23 replied:
All done. Did not go crazy, too early in the morning here right now. :-)

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +139,17 @@ def update_state(self, new_state):
         """
         raise NotImplementedError()
 
+    def call_service(self, domain, service, params, attribute=None):
```
Sorry, missed the "service_data" and thought you wanted attribute instead of value based on IM. Already changed accordingly in my local copy.

ehendrix23 commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +139,17 @@ def update_state(self, new_state):
+            ATTR_ENTITY_ID: self.entity_id,
+            ATTR_DISPLAY_NAME: self.display_name,
+            ATTR_SERVICE: service,
+            ATTR_VALUE: attribute
```
Wasn't sure what was considered best practice here, I do like that a lot better as well. Changed in local copy.

ehendrix23 commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +288,29 @@ def humanify(hass, events):
+                entity_id = data.get(ATTR_ENTITY_ID)
+                dsp_name = data.get(ATTR_DISPLAY_NAME) or entity_id
+                service = data.get(ATTR_SERVICE) or ''
+                attribute = data.get(ATTR_VALUE)
```
Already updated in local copy.

ehendrix23 commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +288,29 @@ def humanify(hass, events):
+                service = data.get(ATTR_SERVICE) or ''
+                attribute = data.get(ATTR_VALUE)
+
+                attr_msg = " to {}".format(attribute) if attribute else ''
```
Already updated :-)

ehendrix23 commented:
@cdce8p I've already addressed all items in my local copy except the one where I check for the values. Awaiting your final thoughts on that one and will then update accordingly. :-)

cdce8p commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +288,29 @@ def humanify(hass, events):
+                    'when': event.time_fired,
+                    'name': 'HomeKit',
+                    'message': message,
+                    'domain': 'homekit',
```
Can you also import `'homekit'` from `homekit.const` with `from .. import DOMAIN as DOMAIN_HOMEKIT`?

ehendrix23 replied:
Done, should it be DOMAIN_HOMEKIT or HOMEKIT_DOMAIN? Asking as in logbook there is an import DOMAIN AS HA_DOMAIN from homeassistant.const.

cdce8p replied:
Doesn't really matter, as long as it is unique.

ehendrix23 commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +288,29 @@ def humanify(hass, events):
+                data = event.data
+                entity_id = data.get(ATTR_ENTITY_ID)
+                dsp_name = data.get(ATTR_DISPLAY_NAME) or entity_id
+                service = data.get(ATTR_SERVICE) or ''
```
Removed

ehendrix23 commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +288,29 @@ def humanify(hass, events):
+                attribute = data.get(ATTR_VALUE)
+
+                attr_msg = " to {}".format(attribute) if attribute else ''
+                dsp_msg = " for {}".format(dsp_name) if dsp_name else ''
```
Done :-)

ehendrix23 committed:
Cleaunup in logbook

Missed renaming of parameter in accessories => fixed.
Some cleanup updates in logbook streamlining it based that this is a HomeKit only event and thus expectation is that required values will be passed.

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -137,6 +139,18 @@ def update_state(self, new_state):
+            ATTR_VALUE: value
+        }
+        self.hass.bus.fire(EVENT_HOMEKIT_CHANGED, event_data)
+
```
Probably super unnecessary, but can you move the empty line from here between `event_data` and `bus.fire`?

ehendrix23 replied:
Done. :-)

cdce8p commented on homeassistant/components/logbook.py:
```diff
@@ -286,6 +289,26 @@ def humanify(hass, events):
+                entity_id = data[ATTR_ENTITY_ID]
+
+                value_msg = " to {}".format(
+                    data.get[ATTR_VALUE]) if data.get[ATTR_VALUE] else ''
```
I think that I might have mislead you here. Sorry for that.
For value it is probably better to define it beforehand:
```py
value = data.get('ATTR_VALUE`)

value_msg = " to {}".format(value) if value else ''
```

For `ATTR_SERVICE` and `ATTR_DISPLAY_NAME` below the formatting is correct.

Another point probably also unnecessary, since we'll have use `get` for `value` can you use it for `entity_id` as well? Just looks a bit better :)

The empty line between `value_msg` and `message` would be unnecessary.

ehendrix23 replied:
Done :-)

cdce8p reviewed (commented):
Just two quick comments.
After these I think we only need to update the tests, expand to all other types and we are done.

ehendrix23 committed:
Last cleanup :-)

Added variable value for data.get(ATTR_VALUE)
Added empty line before bus.fire and removed after bus.fire

ehendrix23 commented:
Both those are done.

If we're good then I'll go ahead and update all other ones to call call_service for state changes.
For updating the tests I would really need help and guidance :-) (Like I have not received any up till now <g>).

cdce8p commented:
Good work so far!

When you change the other types, make sure to rebase your branch. I'm not 100% sure there haven't been any updates: 
https://developers.home-assistant.io/docs/en/development_catching_up.html
That way you're on the save side and we don't miss anything.

As for tests: I'll take a look at them later and get back to you.

ehendrix23 committed:
Added logging to logbook for all other types

Added the calling of self.call_service instead of self.hass.services.call for all other types.

ehendrix23 commented:
@cdce8p Well, that rebase did not work out as it should. Made a mistake and resulted in a whole bunch of other commit being part of my branch and part of my commits.
I think I got it resolved now. :-)

Updated all the other types for HomeKit and are part of this last commit.

cdce8p committed:
Small changes

cdce8p committed:
Added tests

* Logbook test
* HomeAccessory.call_service test
* Updated homekit/test_type_cover.py

cdce8p committed:
Convert call_service to async

cdce8p commented:
I went ahead and pushed some changes to your branch. Take a look at the last commits.
Mainly I converted `call_service` to async and started with the tests.

Take a look at my changes to `test_type_covers.py` to see what to do for the other type tests. Should be pretty strait forward. Except for the remaining type ones, all tests should be finished.

Before you start working on it, make sure to fetch your remote branch first and then rebase your local branch onto it, so you have my commits as well.

ehendrix23 committed:
Enhanced lock test to confirm logbook event.

Enhanced lock test for logbook event from locks.

cdce8p commented on tests/components/homekit/test_type_locks.py:
```diff
@@ -1,6 +1,8 @@
 import pytest
 
+from homeassistant.components.homekit.const import (
+    ATTR_VALUE, ATTR_DISPLAY_NAME)
```
Since you only need `ATTR_VALUE`, one line is completely enough here:
`from homeassistant.components.homekit.const import ATTR_VALUE`

ehendrix23 committed:
Added all other tests

Add tests for all other types.
Fixed issue for type_thermostat where entry in logbook would always state ºC instead of the actual measurement (ºC or ºF).

ehendrix23 committed:
Fixed import of ATTR_VALUE

Changed:
from homeassistant.components.homekit.const import (
    ATTR_VALUE)
to
from homeassistant.components.homekit.const import ATTR_VALUE

ehendrix23 committed:
Fixed import issue in test_type_locks

Fixed the import issue in test_type_locks.

cdce8p commented:
Looks good to me. Great work you did here :1st_place_medal:
